### PR TITLE
riscv-elf.md: add description of the code models

### PR DESCRIPTION
**Pull request by ebahapo**
Add a brief descritption of the existing code models.

### CODE REVIEW

File: riscv-elf.md [hunk 1]
```diff
@@ -320,6 +321,31 @@ default ABIs for specific architectures:
...
+memory position using a literal as its argument.  Rather, instructions exist
+that, when combined together, can then be used to refer to a memory position
+via its literal.  And, when not, other data structures are used to help the
+code to address the memory space.  The coding conventions governing their use
+are known as code models.
+
+### Medium low
+
+The medium low code model, or `medlow`, allows the code to address the lower
+2 GiB of the address space.  By using the instruction `lui` and `ld` or `addi`
```
Comment (jim-wilson) [inline riscv-elf.md:337]: This says lui and ld, but the medany description says auipc and st.  Not clear if it is useful or confusing for the ld/st difference here.

File: riscv-elf.md [hunk 2]
```diff
@@ -320,6 +321,31 @@ default ABIs for specific architectures:
...
+address space.  There is no single instruction that can refer to an arbitrary 
+memory position using a literal as its argument.  Rather, instructions exist
+that, when combined together, can then be used to refer to a memory position
+via its literal.  And, when not, other data structures are used to help the
+code to address the memory space.  The coding conventions governing their use
+are known as code models.
+
+### Medium low
+
+The medium low code model, or `medlow`, allows the code to address the lower
```
Comment (jim-wilson) [inline riscv-elf.md:336]: This says lower 2GiB.  The medany description says +/- 2GiB which is inconsistent.  The med low case is actually low 2GiB and upper 2GiB except that I don't know of any hardware that supports negative addresses in rv64, which would require 64 address lines, so for rv64 it is only the low 2GiB we support, where as for rv32 we support the entire 32-bit address range.  Not clear if it helps or confuses to go into the details though.

File: riscv-elf.md [hunk 3]
```diff
@@ -320,6 +321,31 @@ default ABIs for specific architectures:
...
+2 GiB of the address space.  By using the instruction `lui` and `ld` or `addi`
+instructions, for example, a 32-bit address literal can be produced.
+This code model is not position independent.
+
+### Medium any
+
+The medium any code model, or `medany`, allows the code to address the range
+between -2 GiB and +2 GiB from its position.  By using the instruction `auipc`
```
Comment (jim-wilson) [inline riscv-elf.md:344]: While the instructions do support +/-2GiB.  The code model itself only supports 2GiB objects.  We need to support the case where the first byte references the last byte, and the last byte references the first byte, and hence we can't support in general executables larger than 2GiB total.  Anything larger than 2GiB might fail with out of range reloc errors and hence isn't officially supported.  Again, I'm not sure whether the details help or confuse.

**Comment by lenary** [comment]
It would be great to:
- note the LLVM names for these (small, medium)
- show the function calling sequences explicitly

File: riscv-elf.md (hunk 1 above)
Comment (ebahapo) [reply riscv-elf.md:337]: This is about `medlow` here...

**Comment by ebahapo** [comment]
> It would be great to:
> 
> * note the LLVM names for these (small, medium)
> * show the function calling sequences explicitly

Can you please point to where these remarks apply?

File: riscv-elf.md [hunk 4]
```diff
@@ -320,6 +321,31 @@ default ABIs for specific architectures:
...
+
+The RISC-V architecture constrains the addressing of positions in the
+address space.  There is no single instruction that can refer to an arbitrary 
+memory position using a literal as its argument.  Rather, instructions exist
+that, when combined together, can then be used to refer to a memory position
+via its literal.  And, when not, other data structures are used to help the
+code to address the memory space.  The coding conventions governing their use
+are known as code models.
+
+### Medium low
```
Comment (lenary) [inline riscv-elf.md:334]: The name for this in clang is 'small' (though clang accepts 'medlow').

File: riscv-elf.md [hunk 5]
```diff
@@ -320,6 +321,31 @@ default ABIs for specific architectures:
...
+
+The medium low code model, or `medlow`, allows the code to address the lower
+2 GiB of the address space.  By using the instruction `lui` and `ld` or `addi`
+instructions, for example, a 32-bit address literal can be produced.
+This code model is not position independent.
+
+### Medium any
```
Comment (lenary) [inline riscv-elf.md:341]: The name for this in clang is 'medium' (though clang accepts 'medany').

File: riscv-elf.md (hunk 3 above)
Comment (ebahapo) [reply riscv-elf.md:344]: Methinks that this is implied in the description above, albeit perhaps too implicitly.

File: riscv-elf.md (hunk 1 above)
Comment (jim-wilson) [reply riscv-elf.md]: I was trying to make a point that one example uses ld and the other example uses st.  It isn't clear why they are different, and whether that is irrelevant or potentially confusing. But I think this is a very minor point.

Comment (jim-wilson) [reply riscv-elf.md]: Oh, I see you fixed it.  Nevermind.

File: riscv-elf.md (hunk 3 above)
Comment (jim-wilson) [reply riscv-elf.md]: Yes, I think it is a bit too implicit, but it is a minor point.

**Comment by ebahapo** [comment]
Ping! 🛎

**Pull request merged**